feat(bb): Introduce chunks for univariate computation for the AVM

jeanmon opened the pull request:
This PR introduces a more even vertically distributed trace chunks processing among threads in the univariate computation as part of sumcheck. This leads to a substantial speed up of sumcheck.

Let `t` be the number of threads. The processing of the rows of a circuit used to be

* thread_1 (`round_size/t` rows)
* ...
* thread_t (`round_size/t` rows) [end of circuit]

while this PR introduces the possibility of having the processing be interleaved, and therefore the load is more uniformly balanced across threads:

* thread_1 (`chunk_thread_portion_size` rows)
* ...
* thread_t (`chunk_thread_portion_size` rows)
* thread_1 (`chunk_thread_portion_size` rows)
* ...
* thread_t (`chunk_thread_portion_size` rows)
* ...
* [end of circuit]

This PR improves #12703 measurement from 8.5 seconds to 2.4 seconds.

Commit:
Introduce chunks for univariate computation for the AVM

fcarreiro commented:
The gains look amazing! but... I think we should wait until we have the capacity to do a full trace in VM2 for this. Also I think think 
1. the current approach couples the sumcheck_round with the concept of avm flavor, and requires ifdefs, which kind of pollutes the code a lot. I think we can do better.
2. we should have a more detailed PR explanation with tracy graphs (before/after) on the threads, explaining why this approach works both for short and fuller traces

Commit:
Removing dependency of sumcheck to avm

jeanmon commented:
@fcarreiro I addressed 1) as it was indeed ugly.

fcarreiro committed:
rollback last change

fcarreiro committed:
use concepts

fcarreiro committed:
fix concept

fcarreiro commented:
> The gains look amazing! but... I think we should wait until we have the capacity to do a full trace in VM2 for this. Also I think think
> 
> 1. the current approach couples the sumcheck_round with the concept of avm flavor, and requires ifdefs, which kind of pollutes the code a lot. I think we can do better.
> 2. we should have a more detailed PR explanation with tracy graphs (before/after) on the threads, explaining why this approach works both for short and fuller traces

Pushed some changes to fix 1

fcarreiro commented on barretenberg/cpp/src/barretenberg/sumcheck/sumcheck_round.hpp:
```diff
@@ -171,19 +193,21 @@ template <typename Flavor> class SumcheckProverRound {
-                                                extended_edges,
-                                                relation_parameters,
-                                                gate_sparators[(edge_idx >> 1) * gate_sparators.periodicity]);
+            for (size_t chunk_idx = 0; chunk_idx < num_of_chunks; chunk_idx++) {
```
It feels like it's not the first time I've seen this kind of chunking in BB. We should probably (eventually) track down the uses and write a variant of `parallel_for` that does it for you.

fcarreiro commented:
Having fixed (1) I'm not against merging this as long as crypto is ok with it. We should however revisit it later.

Commit:
Addressing review feedback

fcarreiro commented on barretenberg/cpp/src/barretenberg/vm2/generated/flavor.hpp:
```diff
@@ -95,6 +95,14 @@ class AvmFlavor {
+    // greater or equal to 2.
+    static constexpr size_t MAX_CHUNK_THREAD_PORTION_SIZE = 32;
+    static_assert(MAX_CHUNK_THREAD_PORTION_SIZE >= 2);
+    static_assert((MAX_CHUNK_THREAD_PORTION_SIZE & (MAX_CHUNK_THREAD_PORTION_SIZE - 1)) == 0);
```
This is done at compile time so probably %2 is clearer :)

(also in the hbs)

jeanmon replied:
@fcarreiro This is not %2 but a test for a power of 2.

fcarreiro replied:
oh oh my bad

fcarreiro commented on barretenberg/cpp/src/barretenberg/vm2/generated/flavor.hpp:
```diff
@@ -95,6 +95,14 @@ class AvmFlavor {
+    // that a given thread will process per chunk. This constant is assumed to be a power of 2
+    // greater or equal to 2.
+    static constexpr size_t MAX_CHUNK_THREAD_PORTION_SIZE = 32;
+    static_assert(MAX_CHUNK_THREAD_PORTION_SIZE >= 2);
```
I think you are assuming >= 2 and %2 in the sumcheck code, so that would be the place to put these two asserts, which would then run for any Flavor that specifies MAX_CHUNK_THREAD_PORTION_SIZE.

(try it out first, because IDK how the `if constexpr` plays with the static_assert)

jeanmon replied:
Works and I pushed a fix.

fcarreiro commented:
Actually now it looks good to merge!

Commit:
Further adressing of review feedback

lucasxia01 commented on barretenberg/cpp/src/barretenberg/vm2/generated/flavor.hpp:
```diff
@@ -95,6 +95,12 @@ class AvmFlavor {
+    // evenly processed by all the threads. This constant defines the maximum number of rows
+    // that a given thread will process per chunk. This constant is assumed to be a power of 2
+    // greater or equal to 2.
+    static constexpr size_t MAX_CHUNK_THREAD_PORTION_SIZE = 32;
```
why 32?

lucasxia01 commented on barretenberg/cpp/src/barretenberg/sumcheck/sumcheck_round.hpp:
```diff
@@ -160,8 +165,34 @@ template <typename Flavor> class SumcheckProverRound {
+
+            // When the trace is shrunk to a point where the chunk portion size per thread is lower than 2,
+            // we fall back to a single chunk, i.e., we keep the "non-AVM" values.
+            if (thread_portion_size_candidate >= 2) {
```
why is there an if here? Seems unnecessary?

jeanmon replied:
I added more explanations and actually could simplify a bit the logic. See new version.

lucasxia01 commented on barretenberg/cpp/src/barretenberg/sumcheck/sumcheck_round.hpp:
```diff
@@ -160,8 +165,34 @@ template <typename Flavor> class SumcheckProverRound {
+            static_assert(Flavor::MAX_CHUNK_THREAD_PORTION_SIZE >= 2);
+            static_assert((Flavor::MAX_CHUNK_THREAD_PORTION_SIZE & (Flavor::MAX_CHUNK_THREAD_PORTION_SIZE - 1)) == 0);
+
+            const auto thread_portion_size_candidate =
```
naming is not great.. can't think of proper naming at the moment but at least requires a comment to what this is

jeanmon replied:
yeah, not so easy to have great terminology. I did no change it but added plenty of explanations.

lucasxia01 commented on barretenberg/cpp/src/barretenberg/sumcheck/sumcheck_round.hpp:
```diff
@@ -160,8 +165,34 @@ template <typename Flavor> class SumcheckProverRound {
         size_t num_threads = bb::calculate_num_threads_pow2(round_size, min_iterations_per_thread);
-        size_t iterations_per_thread = round_size / num_threads; // actual iterations per thread
 
+        // In the AVM, the trace is more dense at the top and therefore it is worth to split the work over the threads
```
this section just has a lot of logic thats hard to follow. All of the divisions and unclear names make it hard to parse.

lucasxia01 commented on barretenberg/cpp/src/barretenberg/sumcheck/sumcheck_round.hpp:
```diff
@@ -160,8 +165,34 @@ template <typename Flavor> class SumcheckProverRound {
-        size_t iterations_per_thread = round_size / num_threads; // actual iterations per thread
 
+        // In the AVM, the trace is more dense at the top and therefore it is worth to split the work over the threads
+        // a bit more evenly on the vertical axis. To achieve this, we split the trace into chunks and each thread
```
this comment could have more detail in terms what you mean by "more evenly on the vertical axis"

lucasxia01 reviewed (commented):
Looks fine, and the results are great! I'm not sure how well tested this code is - some of the logic is a little tricky so I would hope for some better testing. Also feel like it needs to be more readable.

Commit:
Addressing review feedback

jeanmon commented:
@lucasxia01 I added a significant number of explanations and explained the required properties to be satisfied and added a little proof of why the code satisfies this. I hope this gives enough confidence that the code is correct. it is a bit hard to unit test this. In any case, I think there is enough safeguard that the new code does not affect non-AVM parts in any way.

ledwards2225 commented:
@jeanmon This is more or less exactly what I had in mind with [this](https://github.com/AztecProtocol/barretenberg/issues/1195) issue. @lucasxia01 do you see any reason why this same mechanism isn't applicable for us in the PG context?

fcarreiro commented:
Personally I think this approach is noticeably better if you don't have uniform density (like in our case, we get many times improvement), and even if you do, it should work at least as good (assuming poor cache locality, which in any case you could probably improve with a bigger thread chunk; but that needs hard data).

lucasxia01 reviewed (approved):
Thanks for the fantastic comments!
